### ISSUE TITLE
Does this actually work?

### ISSUE THREAD

**Issue by tsengalb99** [Azure/MS-AMP#178]
I tried replicating your results on nanoGPT and with your MS-AMP Examples repo and couldn't. Now I may have done something wrong, but this paper https://arxiv.org/abs/2405.18710 gets basically the same thing I'm getting so I'm inclined to believe the public code does not work out of the box. Can you clarify on whether the results in the FP8 LM paper are reproducible or not? For example, does Figure 8 use the same training configuration for all curves or different hyperparameters? Can you detail how you produced that figure? Thanks

**Comment by wkcn** [comment]
Hi @tsengalb99 , thanks for your attention to our work!

Regarding to your questions,

>  Can you clarify on whether the results in the FP8 LM paper are reproducible or not?

The results are reproducible. The codes related to our paper are open-souced and verified.

> Does Figure 8 use the same training configuration for all curves or different hyperparameters?

The same training configuration except of the data types in the optimizer. They are corresponding to Table 6.

> Can you detail how you produced that figure?

You can refer to the scripts: https://github.com/Azure/MS-AMP-Examples/tree/main/gpt3

- Compute
BF16: disable MS-AMP
FP8: enable MS-AMP

- Comm
The weight gradient is a FP8 tensor with a scaling factor.
https://github.com/Azure/MS-AMP/blob/0a2cd721fa68ee725e3b2fb132df02ddb8069d62/msamp/megatron/optimizer/distrib_optimizer.py#L796-L804

- Master weight
https://github.com/Azure/MS-AMP/blob/0a2cd721fa68ee725e3b2fb132df02ddb8069d62/msamp/megatron/optimizer/distrib_optimizer.py#L51

- Optimizer States
https://github.com/Azure/MS-AMP/blob/0a2cd721fa68ee725e3b2fb132df02ddb8069d62/msamp/optim/adamw.py#L105-L106
The arguments named `exp_avg_dtype` and `exp_avg_sq_dtype` is corresponding to the datatypes of first-order and second-order gradient moments.

The training curves of MS-AMP FP8 and BF16 should be close. There might be something wrong if not. I will check it.

**Comment by tsengalb99** [comment]
Thanks, I'll take a look at those. Just as a sanity check, how are you able to match BF16 mixed precision using FP32 master weights with FP8 and BF16 master weights when it is widely accepted that pure BF16 training is unable to match BF16/FP32 mixed precision? Even with stochastic rounding, there is usually a gap between pure BF16 training and BF16/FP32 MP (eg https://storage.googleapis.com/deepmind-media/research/language-research/Training%20Gopher.pdf).

**Comment by wkcn** [comment]
@tsengalb99 
In MS-AMP, the master weight is not pure BF16 or pure FP16. It is a FP16 tensor with a scaling factor. The scaling factor extends the range of FP16.

**Comment by tsengalb99** [comment]
Can you elaborate on why scaling lets us use 16 bit master weights? My understanding is that the issue with BF16 training is the update magnitude is much smaller than the parameter magnitude near convergence, so these small values get "lost" during addition. I don't see why scaling solves this, since, at the end of the day, you're still adding two values with different magnitudes. FP16 has more precision than BF16 so this issue is less prevalent with FP16, but Table 6 2b indicates that BF16 master weights are sufficient for matching FP32.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by wkcn** [comment]
Hi @tsengalb99 , I am training nanoGPT-124M with MS-AMP O1.
Here are the training loss and the validation loss. The gap between BF16 and MS-AMP O1 is small.

![image](https://github.com/user-attachments/assets/ac4bf18c-0eb1-44ae-806b-4d8c808a0051)


Source Code: https://github.com/wkcn/nanoGPT
Difference between the branch msamp and the original nanoGPT: https://github.com/wkcn/nanoGPT/commit/3e2572deeb88e63624ab602c70cd63771fe1d434 
MS-AMP: the latest code, v0.4.0

## Reproduce Steps
- BF16
```bash
torchrun --standalone --nproc_per_node=8 train.py config/train_gpt2.py
```

- MS-AMP O1
```bash
torchrun --standalone --nproc_per_node=8 train.py config/train_gpt2_msamp_O1.py
```

It is worth noting that the last linear layer should be an FP32/BF16 Linear layer. It affects the model performance.
https://github.com/wkcn/nanoGPT/blob/3e2572deeb88e63624ab602c70cd63771fe1d434/model.py#L133-L134
![image](https://github.com/user-attachments/assets/635c36c0-221e-4507-b6fd-c4ff667f6d5f)

**Comment by tsengalb99** [comment]
Thanks, I'll take a look at your nanoGPT integration. If your results are with FP16 master weights and scaling then that does seem possible due to having more mantissa bits than BF16.

**Comment by tocean** [comment]
Take-away is the last linear should not be replaced with Fp8Linear for transformer-based model.

**Comment by tsengalb99** [comment]
Yes, that seems to be the case. The other place my code differed is that I didn't know I had to use msamp's clip grad norm function instead of the torch one.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by wkcn** [comment]
@tsengalb99 It will raise an exception that torch clip does not handle MS-AMP's ScalingTensor.
